### ISSUE TITLE
GenericFilter: add convenience API for programmatic filter configuration

### ISSUE THREAD

**Issue by fractal3000** [jmix-framework/jmix#5351]
It is the continue of the issue #5083 

## Description

Creating a filter configuration in Java code requires performing a sequence of initialization steps in a specific order. The correct order is not documented and differs from what happens automatically in XML. Missing a step produces no compile-time or runtime error, but the filter behaves incorrectly: for example, calling `PropertyFilter.setProperty()` before `PropertyFilter.setDataLoader()` causes the operation selector to silently use defaults; not calling `FilterComponent.setConditionModificationDelegated(true)` makes conditions non-removable at runtime.

## Use case

A developer needs to set up a `GenericFilter` configuration in Java — for example, to apply a default configuration with values that depend on runtime context not available at XML load time, or to dynamically add and remove conditions based on user role. Currently this requires reading the XML loader source to discover what initialization steps are needed.

## Expected behavior

Creating a filter component or configuration programmatically should be as reliable as declaring it in XML. The framework should enforce the required initialization order, not leave it to the caller.

## Impact

Without a safe programmatic API, `GenericFilter` is effectively XML-only for anything non-trivial. Developers who need runtime configurations are forced to work with low-level internals.

**Comment by mig-sam** [comment]
Tested on:
Jmix version: 3.0.999-SNAPSHOT
Jmix Studio Plugin Version: 3.0.SNAPSHOT8478-261
IntelliJ version: IntelliJ IDEA 2026.1

**Comment by fractal3000** [comment]
# GenericFilter builder API — QA scenarios (filter-test)

Manual-test checklist for the programmatic `GenericFilter` builder API (#5351). The example app is
attached as a zip archive with the full project code (**filter-test**). Run it
(`./gradlew bootRun`, log in as `admin` / `admin`), then open the **Scenarios** menu. Each *Scenario*
view configures its filter in code (`onInit`); QA only opens the view and verifies the configuration
selector + grid. The **Scenarios — Overview** menu item describes every scenario inside the app.

## Reference data (12 seeded orders)

| status | count | priority | count | active=true | amount ≥ 1000 | customer set | description set |
|---|---|---|---|---|---|---|---|
| OPEN 4 · IN_PROGRESS 3 · CLOSED 3 · CANCELLED 2 | 12 | CRITICAL 3 · HIGH 3 · NORMAL 4 · LOW 2 | 12 | 7 | 5 | 10 | 8 |

OPEN = ORD-001,002,007,009 · CLOSED = 004,005,010 · CRITICAL = 001,007,011 · HIGH = 002,008,010 ·
Acme orders = 001,003,008 · no customer = 006,010.

## Scenarios

| # | View / demonstrates | Check → expected |
|---|---|---|
| 1 | Simple PropertyFilter | Opens with Status=Open prefilled → **4 rows**; Reset keeps Open |
| 2 | JpqlFilter parameter types | Field editors render: void=**checkbox** (checked → active=true, **7**), String=text, enum=combo, BigDecimal=number, LocalDate=date picker |
| 3 | Multiple configurations | "Open Orders" active → **4**; switch to "Closed Orders" → **3** |
| 4 | Deletable vs protected | "Critical…(protected)" active → **3**; in selector protected has **no** Remove (×), "High (deletable)" **has** it; switch → HIGH **3** |
| 5 | Direct RunTimeConfiguration API | "Dynamic Filter" shows one `amount ≥` condition with Remove (×); enter 1000 → **5** |
| 6 | Builder fail-fast diagnostics | Baseline=Open (**4**). Each button shows a caught `IllegalStateException` notification (duplicate id / one-shot reuse / missing input / no DataLoader); grid unaffected |
| 7 | Reference-field filters | (1) customer picker → pick *Acme Corporation* → **3**; (3) name contains "Inc" → **2** |
| 8 | Operation visibility | (1) number prefilled "ORD-001" → **1**; (1) has editable operator dropdown, (2) no operator label, (3) editable + no label |
| 9 | IN-expression filters | Multi-select combos; status IN {Open,Closed} → **7**; priority IN {Critical,High} → **6** |
| 10 | JPQL JOIN clause | email contains "acme" via (1) explicit join and (2) dot-path → both **3** (identical) |
| 11 | Nested paths (root OR) | name contains "Globex" → **3**; *or* email starts "billing" → **3**; either field alone filters (OR) |
| 12 | Mixed XML + Java configs | Selector has 3: ① All orders (XML), ② Active only (XML), ③ Open+amount (Java); ② → active=true **7**; XML configs cannot be removed |
| 13 | removeConfiguration() | "Permanent" active → **4**; *Try remove Permanent* → "NOT removed"; *Remove Critical/High* → gone from selector |
| 14 | setCurrentConfiguration() | A=Open (**4**); →B → HIGH **3**; →C (unregistered) → no-op + log warning; →addConfiguration+setCurrent(C) → active=true **7** |
| 15 | IS_SET (unary) | No value input; description IS SET → **8**; customer IS SET → **10**; amount IS SET → **12** |

## Adding your own filters (new API)

QA can extend the example without deep coding — clone an existing `OrderScenarioNView` or edit its
`onInit`. The pattern:

```java
@ViewComponent
private GenericFilter genericFilter;            // bound to a DataLoader in the view XML

@Subscribe
public void onInit(InitEvent event) {
    var cb = genericFilter.filterComponentBuilder();

    // 1) build a condition
    var byStatus = cb.<OrderStatus>propertyFilter()         // or jpqlFilter() / jpqlFilter(Class) / groupFilter()
            .property("status").operation(PropertyFilter.Operation.EQUAL).build();

    // 2) register it in a configuration
    genericFilter.runtimeConfigurationBuilder()
            .id("myConfig").name("My filter")               // id must be unique within the filter
            .add(byStatus, OrderStatus.OPEN)                // .add(fc) / .addAll(fc...) / .add(fc, defaultValue)
            .makeCurrent()                                  // optional: active on open
            .allowDeletion()                                // optional: user may delete it
            .buildAndRegister();
}
```

Rules to keep in mind:
- The filter **must have a `DataLoader`** (set via `dataLoader="..."` in the view XML) — builders fail
  fast otherwise.
- Builders are **one-shot**: one builder instance = one component / one configuration.
- `jpqlFilter()` (no arg) → a **checkbox** toggling a fixed WHERE; `jpqlFilter(Class)` → a typed input;
  `defaultValue(true)` on the void variant makes it active on open.

The 15 `OrderScenarioNView` classes in the attached project are working templates — copy one as a
starting point for your own filter.


Sample project [filter-test.zip](https://github.com/user-attachments/files/29051629/filter-test.zip)

**Comment by mig-sam** [comment]
1. Scenario 3 — Multiple configurations - failed

`Closed Orders` configuration returns an empty grid, but should return 3 rows
Looks like where clause used from two configurations (for Status = OPEN and Status = CLOSED):
````
eclipselink.logging.sql : SELECT LIMIT ? ? ID AS a1, ... STATUS AS a8, ... CUSTOMER_ID AS a10
  FROM FILTER_TEST_ORDER WHERE ((STATUS = ?) AND (STATUS = ?)) ORDER BY NUMBER_
    bind => [0, 50, O, C]
````

<img width="1575" height="555" alt="Image" src="https://github.com/user-attachments/assets/c005e6aa-1123-401c-93ea-ae72edc2201a" />

**Comment by fractal3000** [comment]
The problem is not reprodused.

**Comment by mig-sam** [comment]
1. update
Problem: Scenario 3 — Multiple configurations 
will be resolved in separate ticket: #5400

**Comment by mig-sam** [comment]
Tested on:
Jmix version: 3.0.999-SNAPSHOT
Jmix Studio Plugin Version: 3.0.SNAPSHOT8507-261
IntelliJ version: IntelliJ IDEA 2026.1
